### PR TITLE
Improvements to artifact-search and some bug fixes.

### PR DESCRIPTION
**Pull request by GoogilyBoogily**

### CODE REVIEW

File: lib/artifact-search.js [hunk 1]
```diff
@@ -68,28 +78,38 @@ module.exports = (bluebird, error, logger, requestOptions, requestPromise, respo
...
+         * @property {(string|string[])} search A string or array of strings
+         *                                      for the search criteria. Can be
+         *                                      undefined.
+         * @property {(string|string[])} sort A string or array of strings for
+         *                                    how the results will be sorted.
+         * @property {number} limit A number denoting how many results should
+         *                          be returned from the search.
+         */
+
+        /**
```
Comment (quantumew) [inline lib/artifact-search.js]: According to the style guide JSDoc blocks should be padded by 2 newlines :-( - "Blank line above comments, two blank lines above JSDoc comments." I only see two instances of this. Not that I really care much.

File: lib/response-handler.js [hunk 2]
```diff
@@ -86,9 +86,18 @@ module.exports = (bluebird, error, logger, parseLinkHeader, URI) => {
...
+            // all onto an array. If we received just one, push just that
+            // one onto the array.
+            if (Array.isArray(links.item)) {
+                links.item.forEach((val) => {
+                    urls.push(val.url);
+                });
+            } else {
+                urls.push(links.url);
+            }
+
```
Comment (quantumew) [inline lib/response-handler.js]: This won't work for a single link, you want this:
```
urls.push(links.item.url);
```
If it is a single link, should we just return a string? So it returns an array for multiple or string for single? What do you think?

File: lib/artifact-search.js (hunk 1 above)
Comment (GoogilyBoogily) [inline lib/artifact-search.js]: Alright, cool. Fixed.

File: lib/artifact-search.js [hunk 3]
```diff
@@ -68,28 +78,38 @@ module.exports = (bluebird, error, logger, requestOptions, requestPromise, respo
...
          */
-        search(search, sort) {
+        search(searchParameters) {
             var options, urls;
 
             logger.debug(`In ${this.constructor.name}.search()`);
+            logger.debug(`Search parameters: ${JSON.stringify(searchParameters, null, 4)}`);
 
-            return validateSearch(search).then(() => {
+            return validateSearch(searchParameters.search).then(() => {
```
Comment (GoogilyBoogily) [inline lib/artifact-search.js]: Got it. Fixed.

File: README.md [hunk 4]
```diff
@@ -30,7 +30,7 @@ Options
 
 The second parameter when requiring the library is an object holding various options that can change the behavior of the library.
 
-* `logLevel` - Sets the desired amount of logging. Values can be: "info", "debug", or "warning". The value defaults to "warning"
+* `logLevel` - Sets the desired amount of logging. Values can be: "info", "debug", or "warning". The value defaults to "warning". WARNING: If the logLevel is set to "debug", the logger will log your Shelf authentication token. Log with "debug" at your own peril.
```
Comment (GoogilyBoogily) [inline README.md]: Yeah, basically too much work and not enough payoff. Also, I made the point where you might wanna make sure if your token is correct in the request object.

File: lib/response-handler.js (hunk 2 above)
Comment (GoogilyBoogily) [inline lib/response-handler.js]: Crap. Yep. Fixed!

File: lib/artifact-search.js [hunk 5]
```diff
@@ -68,29 +78,56 @@ module.exports = (bluebird, error, logger, requestOptions, requestPromise, respo
...
+                }
+
+                if (searchParameters.sort) {
+                    options.body.sort = searchParameters.sort;
+                }
+
+                if (searchParameters.limit) {
+                    options.body.limit = searchParameters.limit;
+                }
+
```
Comment (quantumew) [inline lib/artifact-search.js]: You can just set the searchParameters as the body, right? `options.body = searchParameters` that is essentially what you are doing already. When it is loaded as JSON any undefined properties will just be omitted.

File: lib/artifact-search.js [hunk 6]
```diff
@@ -68,29 +78,56 @@ module.exports = (bluebird, error, logger, requestOptions, requestPromise, respo
 
 
         /**
+         * @typedef {Object} searchParameters
+         * @property {(string|string[])} [search] A string or array of strings
+         *                                        for the search criteria. Can
+         *                                        be undefined (optional).
+         * @property {(string|string[])} [sort] A string or array of strings
```
Comment (quantumew) [inline lib/artifact-search.js]: Sorry didn't catch this originally, I was told array of strings in a JSDoc block is `array.<string>`. Unless that changed, search and sort type should be updated. Did that change?

Comment (GoogilyBoogily) [inline lib/artifact-search.js]: Discussed. Yes.

File: lib/artifact-search.js (hunk 5 above)
Comment (GoogilyBoogily) [inline lib/artifact-search.js]: I... am not a clever man. Fixed.

**Pull request merged**